### PR TITLE
Fix mypy in `utilities.distributed`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?
Part of #7037. This PR disentangles large #7218. It fixes **mypy** typing in `utilities.device_parser.py`.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -216,7 +221,7 @@ def backward(ctx, *grad_output):
 
 
 def all_gather_ddp_if_available(
-    tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False
+    tensor: torch.Tensor, group: Optional[torch.distributed.ProcessGroup] = None, sync_grads: bool = False
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:224]: `torch.distributed.ProcessGroup` is not available for PyTorch <1.7 on Windows

I think writing it in quotes should work

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -239,10 +244,10 @@ def all_gather_ddp_if_available(
 
 
 def register_ddp_comm_hook(
-    model: DistributedDataParallel,
+    model: Type[DistributedDataParallel],
```
Comment (awaelchli) [inline pytorch_lightning/utilities/distributed.py:247]: I don't understand this change. The code shows we are passing in an instance so I would expect the previous type hint was correct.

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -31,21 +31,22 @@
...
 
 
-def rank_zero_only(fn):
+def rank_zero_only(fn: Callable) -> Callable:
     @wraps(fn)
-    def wrapped_fn(*args, **kwargs):
+    def wrapped_fn(*args: Any, **kwargs: Any) -> Optional[Any]:
         if rank_zero_only.rank == 0:
             return fn(*args, **kwargs)
+        return None
```
Comment (awaelchli) [inline pytorch_lightning/utilities/distributed.py:49]: why is this necessary? can mypy not infer that?

Comment (stancld) [reply pytorch_lightning/utilities/distributed.py:49]: If there's no `return None`, mypy raises "Missing return statements". I think `return None` follows the PEP8 recommendations.

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (stancld) [reply pytorch_lightning/utilities/distributed.py:247]: In the previous case, mypy throws the following error
```
pytorch_lightning/utilities/distributed.py:343: error: "Tensor" not callable  [operator]
```
when `model.register_comm_hook(state=ddp_comm_state, hook=ddp_comm_hook)` being called.

However, I agree with you that the previous type hint looks correct. Maybe should we ignore mypy here?

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (stancld) [reply pytorch_lightning/utilities/distributed.py]: Thanks, looks it worked :)

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/distributed.py:247]: There must be a bug somewhere in the typing where `model.register_comm_hook` becomes a tensor :( 
But in any case with the new type annotation, we get: 
`pytorch_lightning/utilities/distributed.py:343: error: "Type[DistributedDataParallel]" has no attribute "register_comm_hook"`

so we should not do it and just leave as is imo

Comment (ananthsub) [reply pytorch_lightning/utilities/distributed.py:247]: @SciPioneer have you seen this typing issue come up elsewhere for DDP comm hooks?

Comment (stancld) [reply pytorch_lightning/utilities/distributed.py]: Hmmm... I don't know why mypy throws an error on my local while it's okay here (gonna check my versions). But bottom line is, mypy typing should be fine for now :)

**Review by awaelchli — APPROVED** [review]
